src/controller.cpp

```diff
@@ -88,23 +83,21 @@ void Controller::setPlan(const nav_msgs::msg::Path & path)
 
 void Controller::setSpeedLimit(const double & speed_limit, const bool & percentage)
 {
-  utils::ControlConstraints constraints;
+  models::ControlConstraints constraints = optimizer_.getDefaultControlConstraints();
 
-  if (speed_limit == nav2_costmap_2d::NO_SPEED_LIMIT) {
-    // Restore default value
-    constraints = utils::ControlConstraints{base_x_vel_, base_y_vel_, base_theta_vel_};
-  } else {
+  if (speed_limit != nav2_costmap_2d::NO_SPEED_LIMIT) {
     if (percentage) {
       // Speed limit is expressed in % from maximum speed of robot
-      constraints.vx = base_x_vel_ * speed_limit / 100.0;
-      constraints.vy = base_y_vel_ * speed_limit / 100.0;
-      constraints.vw = base_theta_vel_ * speed_limit / 100.0;
+      double ratio = speed_limit / 100.0;
+      constraints.vx *= ratio;
```

Comment:
This is wrong as well, this will apply the change on the previously applied change, this must be done using the base values